core/services/synchronization/telemetry_ingress_batch_client.go

```diff
@@ -0,0 +1,301 @@
+	Start() error
+	Close() error
+	Send(TelemPayload)
+}
+
+// NoopTelemetryIngressBatchClient is a no-op interface for TelemetryIngressBatchClient
+type NoopTelemetryIngressBatchClient struct{}
+
+// Start is a no-op
+func (NoopTelemetryIngressBatchClient) Start() error { return nil }
+
+// Close is a no-op
+func (NoopTelemetryIngressBatchClient) Close() error { return nil }
+
+// Send is a no-op
+func (NoopTelemetryIngressBatchClient) Send(TelemPayload) {}
+
+// Healthy is a no-op
+func (NoopTelemetryIngressBatchClient) Healthy() error { return nil }
+
+// Ready is a no-op
+func (NoopTelemetryIngressBatchClient) Ready() error { return nil }
+
+type telemetryIngressBatchClient struct {
+	utils.StartStopOnce
+	url             *url.URL
+	ks              keystore.CSA
+	serverPubKeyHex string
+
+	telemClient  telemPb.TelemClient
+	globalLogger logger.Logger
+	logging      bool
+	lggr         logger.Logger
+
+	wgDone sync.WaitGroup
+	chDone chan struct{}
+
+	telemBufferSize   uint
+	telemMaxBatchSize uint
+	telemSendInterval time.Duration
+
+	workers      map[string]*telemetryIngressBatchWorker
+	workersMutex sync.Mutex
+}
+
+// NewTelemetryIngressBatchClient returns a client backed by wsrpc that
+// can send telemetry to the telemetry ingress server
+func NewTelemetryIngressBatchClient(url *url.URL, serverPubKeyHex string, ks keystore.CSA, logging bool, lggr logger.Logger, telemBufferSize uint, telemMaxBatchSize uint, telemSendInterval time.Duration) TelemetryIngressBatchClient {
+	return &telemetryIngressBatchClient{
+		telemBufferSize:   telemBufferSize,
+		telemMaxBatchSize: telemMaxBatchSize,
+		telemSendInterval: telemSendInterval,
+		url:               url,
+		ks:                ks,
+		serverPubKeyHex:   serverPubKeyHex,
+		globalLogger:      lggr,
+		logging:           logging,
+		lggr:              lggr.Named("TelemetryIngressBatchClient"),
+		chDone:            make(chan struct{}),
+		workers:           make(map[string]*telemetryIngressBatchWorker),
+	}
+}
+
+// Start connects the wsrpc client to the telemetry ingress server
+func (tc *telemetryIngressBatchClient) Start() error {
+	return tc.StartOnce("TelemetryIngressBatchClient", func() error {
+		privkey, err := tc.getCSAPrivateKey()
+		if err != nil {
+			return err
+		}
+
+		tc.connect(privkey)
+
+		return nil
+	})
+}
+
+// Close disconnects the wsrpc client from the ingress server and waits for all workers to exit
+func (tc *telemetryIngressBatchClient) Close() error {
+	return tc.StopOnce("TelemetryIngressBatchClient", func() error {
+		close(tc.chDone)
+		tc.wgDone.Wait()
+		return nil
+	})
+}
+
+func (tc *telemetryIngressBatchClient) connect(clientPrivKey []byte) {
+	tc.wgDone.Add(1)
+
+	go func() {
+		defer tc.wgDone.Done()
+
+		serverPubKey := keys.FromHex(tc.serverPubKeyHex)
+
+		conn, err := wsrpc.Dial(tc.url.String(), wsrpc.WithTransportCreds(clientPrivKey, serverPubKey))
+		if err != nil {
+			tc.lggr.Errorf("Error connecting to telemetry ingress server: %v", err)
```

Comment:
Should not we call `connect` synchronously from `Start` and let `Start` return this error? 
Otherwise, should not we re-try to connect in case it failed to connect? From this point it seems that we have this service doing nothing if it failed to connect, until the node restarts.

Reply:
This was taken from the existing telem ingress client. I think we kept it in its own go routine as `Dial` is a blocking operation that would prevent other services from spinning up ([as noted in the connection manager for FMS](https://github.com/smartcontractkit/chainlink/blob/develop/core/services/feeds/connection_manager.go#L74-L81), which also uses wsrpc)

As for retries, I don't think wsrpc returns this error in practice, but instead will return without an error if it can't connect and then continue to retry the connection under-the-hood

I'll confirm the above with @jkongie and circle back. At the very least, I think I should add a few comments to clarify the behavior

Reply:
Yes let's add a clarifying comment and still need to check if `Send()` function will not make things worse if the client is unconnected at the moment.

Reply:
Added a comment clarifying `connect`

If we try to send telemetry before the connection is ready, we get a `connection is not ready` error from wsrpc that's being logged as a warn, so I don't think there's any harm in trying to send before the connection is established

Reply:
@skubakdj is correct. A failed connection will not return an error as it attempts to retry to establish a connection. An error will be returned on any RPC call into wsrpc which looks like it is being handled.

There is an option to make the initial connection a blocking call (i.e. `Dial` blocks until a connection is established) but I don't think that's needed here.

Reply:
Will remove the go func and let `Start` return this error since `Dial` does not block by default :slightly_smiling_face: